ddterm/shell/extension.js

```diff
@@ -140,6 +142,50 @@ class ExtensionDBusInterface {
     }
 }
 
+class DesktopEntry {
+    constructor() {
+        this.target_file = Gio.File.new_for_path(GLib.build_filenamev(
+            [
+                GLib.get_user_data_dir(),
+                'applications',
+                `${APP_ID}.desktop`,
+            ]
+        ));
+        this.source_file = Gio.File.new_for_path(GLib.build_filenamev(
+            [
+                GLib.get_user_data_dir(),
+                'gnome-shell',
+                'extensions',
+                Me.metadata.uuid,
+                'ddterm',
+                `${APP_ID}.desktop`,
+            ]
+        ));
+    }
+
+    _ensure_install_directory_exists() {
+        var install_directory = this.target_file.get_parent();
+        if (!GLib.file_test(install_directory.get_path(), GLib.FileTest.IS_DIR))
+            install_directory.make_directory_with_parents(null);
+    }
+
+    _is_installed() {
+        return GLib.file_test(this.target_file.get_path(), GLib.FileTest.EXISTS);
+    }
+
+    install() {
+        if (!this._is_installed()) {
```

Comment:
Quote from the documentation for [`.exists()`](https://gjs-docs.gnome.org/gio20~2.0/gio.file#method-query_exists):

> Note that in many cases it is [racy to first check for file existence](https://en.wikipedia.org/wiki/Time_of_check_to_time_of_use)
> and then execute something based on the outcome of that, because the
> file might have been created or removed in between the operations. The
> general approach to handling that is to not check, but just do the
> operation and handle the errors as they come.

Maybe it would be easier to handle the error from `.copy()` (checking that the error is `EXISTS`, of course)? Or.. does it actually throw an error when the target file already exists?

Another concern: if a previous version didn't clean up properly, an outdated file will be kept...

Reply:
I'll handle the error as you suggested, but instead of doing nothing if the file already exists I can overwrite the file with `GLib.File.replace()`. Would that solve the issue?

Reply:
https://gjs-docs.gnome.org/gio20~2.0/gio.file#method-copy

> If the flag #G_FILE_COPY_OVERWRITE is specified an already
existing destination file is overwritten.

Won't `.copy()` with `Gio.FileCopyFlags.OVERWRITE` solve all issues with just one function call, without needing any checks?

Reply:
It works every time using this flag. Thanks for pointing me to that!
Implemented on [5744f79](https://github.com/ddterm/gnome-shell-extension-ddterm/pull/297/commits/5744f79d92c0ebe210b0365b715bc11da11ce39f).